File: js/systems/questSystem.js
class QuestSystem {
    constructor(game) {
        this.game = game;
        this.quests = new Map();
        this.activeQuests = new Set();
        this.completedQuests = new Set();
        this.questLog = new QuestLog(this);
        
        // Carregar quests predefinidas
        this.loadQuests();
    }

    loadQuests() {
        // Adicionar quests do jogo
        QuestDatabase.forEach((questData, questId) => {
            this.quests.set(questId, new Quest(questData));
        });
    }

    startQuest(questId) {
        const quest = this.quests.get(questId);
        if (!quest || this.activeQuests.has(questId) || this.completedQuests.has(questId)) {
            return false;
        }

        if (quest.checkPrerequisites(this)) {
            quest.start(this.game);
            this.activeQuests.add(questId);
            this.questLog.update();
            
            // Notificar o jogador
            this.game.ui.showNotification(`Nova missão: ${quest.title}`);
            
            return true;
        }
        
        return false;
    }

    updateQuest(questId, progress) {
        if (!this.activeQuests.has(questId)) return;
        
        const quest = this.quests.get(questId);
        if (quest) {
            quest.updateProgress(progress);
            this.questLog.update();
            
            if (quest.isComplete()) {
                this.completeQuest(questId);
            }
        }
    }

    completeQuest(questId) {
        const quest = this.quests.get(questId);
        if (!quest || !this.activeQuests.has(questId)) return;
        
        this.activeQuests.delete(questId);
        this.completedQuests.add(questId);
        
        // Entregar recompensas
        quest.giveRewards(this.game.player);
        
        // Atualizar UI
        this.questLog.update();
        
        // Notificar o jogador
        this.game.ui.showNotification(`Missão completada: ${quest.title}`);
        
        // Verificar quests dependentes
        this.checkDependentQuests(questId);
    }

    failQuest(questId) {
        const quest = this.quests.get(questId);
        if (!quest || !this.activeQuests.has(questId)) return;
        
        this.activeQuests.delete(questId);
        quest.fail();
        this.questLog.update();
        
        // Notificar o jogador
        this.game.ui.showNotification(`Missão falhou: ${quest.title}`, 'error');
    }

    checkDependentQuests(completedQuestId) {
        this.quests.forEach((quest, questId) => {
            if (!this.activeQuests.has(questId) && !this.completedQuests.has(questId)) {
                if (quest.checkPrerequisites(this)) {
                    // Automaticamente iniciar quests que dependiam desta
                    if (quest.autoStart) {
                        this.startQuest(questId);
                    }
                }
            }
        });
    }

    isQuestActive(questId) {
        return this.activeQuests.has(questId);
    }

    isQuestCompleted(questId) {
        return this.completedQuests.has(questId);
    }

    getQuestProgress(questId) {
        const quest = this.quests.get(questId);
        return quest ? quest.progress : null;
    }

    getActiveQuests() {
        return Array.from(this.activeQuests).map(id => this.quests.get(id));
    }

    getCompletedQuests() {
        return Array.from(this.completedQuests).map(id => this.quests.get(id));
    }

    reset() {
        this.activeQuests.clear();
        this.completedQuests.clear();
        this.quests.forEach(quest => quest.reset());
        this.questLog.update();
    }

    save() {
        return {
            activeQuests: Array.from(this.activeQuests),
            completedQuests: Array.from(this.completedQuests),
            questProgress: Array.from(this.quests.entries()).map(([id, quest]) => ({
                id,
                progress: quest.progress
            }))
        };
    }

    load(data) {
        this.reset();
        
        data.activeQuests.forEach(id => {
            const quest = this.quests.get(id);
            if (quest) {
                this.activeQuests.add(id);
            }
        });
        
        data.completedQuests.forEach(id => {
            this.completedQuests.add(id);
        });
        
        data.questProgress.forEach(({id, progress}) => {
            const quest = this.quests.get(id);
            if (quest) {
                quest.progress = progress;
            }
        });
        
        this.questLog.update();
    }
}

class Quest {
    constructor(data) {
        this.id = data.id;
        this.title = data.title;
        this.description = data.description;
        this.objectives = data.objectives;
        this.rewards = data.rewards;
        this.prerequisites = data.prerequisites || [];
        this.autoStart = data.autoStart || false;
        this.type = data.type || 'side'; // main, side, daily, etc.
        this.level = data.level || 1;
        
        this.progress = {
            started: false,
            completed: false,
            failed: false,
            objectives: this.objectives.map(obj => ({
                description: obj.description,
                current: 0,
                required: obj.required,
                completed: false
            }))
        };
        
        // Callbacks
        this.onStart = data.onStart;
        this.onComplete = data.onComplete;
        this.onFail = data.onFail;
        this.onProgress = data.onProgress;
    }

    start(game) {
        this.progress.started = true;
        if (this.onStart) {
            this.onStart(game);
        }
    }

    updateProgress(objectiveIndex, amount = 1) {
        const objective = this.progress.objectives[objectiveIndex];
        if (!objective || objective.completed) return;
        
        objective.current = Math.min(objective.current + amount, objective.required);
        objective.completed = objective.current >= objective.required;
        
        if (this.onProgress) {
            this.onProgress(objectiveIndex, objective.current);
        }
    }

    isComplete() {
        return this.progress.objectives.every(obj => obj.completed);
    }

    giveRewards(player) {
        this.rewards.forEach(reward => {
            switch (reward.type) {
                case 'gold':
                    player.addGold(reward.amount);
                    break;
                case 'exp':
                    player.addExperience(reward.amount);
                    break;
                case 'item':
                    player.inventory.addItem(reward.item, reward.amount);
                    break;
                case 'reputation':
                    player.addReputation(reward.faction, reward.amount);
                    break;
            }
        });
        
        if (this.onComplete) {
            this.onComplete(player);
        }
    }

    fail() {
        this.progress.failed = true;
        if (this.onFail) {
            this.onFail();
        }
    }

    reset() {
        this.progress = {
            started: false,
            completed: false,
            failed: false,
            objectives: this.objectives.map(obj => ({
                description: obj.description,
                current: 0,
                required: obj.required,
                completed: false
            }))
        };
    }

    checkPrerequisites(questSystem) {
        return this.prerequisites.every(prereq => {
            switch (prereq.type) {
                case 'quest':
                    return questSystem.isQuestCompleted(prereq.id);
                case 'level':
                    return questSystem.game.player.level >= prereq.level;
                case 'reputation':
                    return questSystem.game.player.getReputation(prereq.faction) >= prereq.amount;
                default:
                    return false;
            }
        });
    }
}

class QuestLog {
    constructor(questSystem) {
        this.questSystem = questSystem;
        this.createUI();
    }

    createUI() {
        // Criar interface do quest log
        this.container = document.createElement('div');
        this.container.className = 'quest-log';
        this.container.style.cssText = `
            position: fixed;
            right: -400px;
            top: 50%;
            transform: translateY(-50%);
            width: 380px;
            max-height: 80vh;
            background: linear-gradient(to bottom, #2c1810, #1a0f0a);
            border: 5px solid #c0a080;
            border-radius: 15px;
            padding: 20px;
            transition: right 0.3s;
            overflow-y: auto;
            color: #e8d5b5;
            font-family: 'MedievalFont', serif;
            z-index: 900;
        `;

        // Título
        const title = document.createElement('h2');
        title.textContent = 'Missões';
        title.style.cssText = `
            color: #daa520;
            text-align: center;
            margin-bottom: 20px;
            border-bottom: 2px solid #c0a080;
            padding-bottom: 10px;
        `;

        // Lista de quests
        this.questList = document.createElement('div');
        this.questList.className = 'quest-list';

        this.container.appendChild(title);
        this.container.appendChild(this.questList);
        document.body.appendChild(this.container);
    }

    update() {
        this.questList.innerHTML = '';
        
        // Agrupar quests por tipo
        const questGroups = {
            main: [],
            side: [],
            daily: []
        };
        
        // Adicionar quests ativas
        this.questSystem.getActiveQuests().forEach(quest => {
            questGroups[quest.type].push(this.createQuestElement(quest));
        });
        
        // Adicionar grupos ao log
        Object.entries(questGroups).forEach(([type, quests]) => {
            if (quests.length > 0) {
                const groupTitle = document.createElement('h3');
                groupTitle.textContent = this.getGroupTitle(type);
                groupTitle.style.color = this.getGroupColor(type);
                this.questList.appendChild(groupTitle);
                
                quests.forEach(questElement => {
                    this.questList.appendChild(questElement);
                });
            }
        });
    }

    createQuestElement(quest) {
        const element = document.createElement('div');
        element.className = 'quest-entry';
        element.style.cssText = `
            margin-bottom: 15px;
            padding: 10px;
            background: rgba(0, 0, 0, 0.3);
            border-radius: 8px;
        `;

        const title = document.createElement('h4');
        title.textContent = quest.title;
        title.style.color = '#daa520';

        const description = document.createElement('p');
        description.textContent = quest.description;
        description.style.fontSize = '14px';

        const objectives = document.createElement('ul');
        objectives.style.listStyle = 'none';
        objectives.style.padding = '0';
        
        quest.progress.objectives.forEach(obj => {
            const li = document.createElement('li');
            li.style.cssText = `
                margin: 5px 0;
                display: flex;
                justify-content: space-between;
                align-items: center;
            `;
            
            const text = document.createElement('span');
            text.textContent = obj.description;
            
            const progress = document.createElement('span');
            progress.textContent = `${obj.current}/${obj.required}`;
            progress.style.color = obj.completed ? '#4CAF50' : '#daa520';
            
            li.appendChild(text);
            li.appendChild(progress);
            objectives.appendChild(li);
        });

        element.appendChild(title);
        element.appendChild(description);
        element.appendChild(objectives);

        return element;
    }

    getGroupTitle(type) {
        switch (type) {
            case 'main': return 'Missões Principais';
            case 'side': return 'Missões Secundárias';
            case 'daily': return 'Missões Diárias';
            default: return 'Outras Missões';
        }
    }

    getGroupColor(type) {
        switch (type) {
            case 'main': return '#ffd700';
            case 'side': return '#c0c0c0';
            case 'daily': return '#87ceeb';
            default: return '#ffffff';
        }
    }

    show() {
        this.container.style.right = '20px';
    }

    hide() {
        this.container.style.right = '-400px';
    }

    toggle() {
        if (this.container.style.right === '20px') {
            this.hide();
        } else {
            this.show();
        }
    }
}

// Banco de Dados de Quests
const QuestDatabase = new Map([
    ['tutorial_quest', {
        id: 'tutorial_quest',
        title: 'Primeiros Passos',
        description: 'Aprenda o básico sobre como jogar.',
        type: 'main',
        level: 1,
        autoStart: true,
        objectives: [
            {
                description: 'Fale com o Velho Sábio',
                required: 1
            },
            {
                description: 'Equipe sua primeira arma',
                required: 1
            },
            {
                description: 'Derrote um boneco de treino',
                required: 1
            }
        ],
        rewards: [
            { type: 'gold', amount: 100 },
            { type: 'exp', amount: 50 },
            { type: 'item', item: 'basic_potion', amount: 3 }
        ]
    }],
    
    ['crystal_quest', {
        id: 'crystal_quest',
        title: 'O Cristal Perdido',
        description: 'Encontre o primeiro fragmento do Cristal de Eldoria.',
        type: 'main',
        level: 2,
        prerequisites: [
            { type: 'quest', id: 'tutorial_quest' }
        ],
        objectives: [
            {
                description: 'Investigue a torre do Velho Sábio',
                required: 1
            },
            {
                description: 'Derrote os inimigos',
                required: 5
            },
            {
                description: 'Recupere o fragmento do cristal',
                required: 1
            }
        ],
        rewards: [
            { type: 'gold', amount: 500 },
            { type: 'exp', amount: 200 },
            { type: 'item', item: 'crystal_fragment', amount: 1 }
        ]
    }]
]);

File: src/systems/quest_system.py
import json
from typing import Dict, List, Optional, Callable
from enum import Enum
import pygame

class QuestStatus(Enum):
    NOT_STARTED = "not_started"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    FAILED = "failed"

class QuestObjective:
    def __init__(self, description: str, target_amount: int = 1):
        self.description = description
        self.target_amount = target_amount
        self.current_amount = 0
        self.completed = False
        
    def update(self, amount: int = 1) -> bool:
        """Update progress towards objective. Returns True if newly completed."""
        if self.completed:
            return False
            
        self.current_amount = min(self.current_amount + amount, self.target_amount)
        if self.current_amount >= self.target_amount:
            self.completed = True
            return True
        return False
        
    def get_progress(self) -> float:
        """Get progress as a percentage."""
        return self.current_amount / self.target_amount

class Quest:
    def __init__(self, quest_id: str, title: str, description: str):
        self.id = quest_id
        self.title = title
        self.description = description
        self.objectives: Dict[str, QuestObjective] = {}
        self.rewards: Dict[str, int] = {}
        self.status = QuestStatus.NOT_STARTED
        self.on_complete: Optional[Callable] = None
        self.on_fail: Optional[Callable] = None
        
    def add_objective(self, objective_id: str, description: str, target_amount: int = 1):
        """Add a new objective to the quest."""
        self.objectives[objective_id] = QuestObjective(description, target_amount)
        
    def add_reward(self, reward_type: str, amount: int):
        """Add a reward for completing the quest."""
        self.rewards[reward_type] = amount
        
    def update_objective(self, objective_id: str, amount: int = 1) -> bool:
        """Update progress of an objective. Returns True if quest is completed."""
        if objective_id not in self.objectives:
            return False
            
        if self.objectives[objective_id].update(amount):
            # Check if all objectives are complete
            if all(obj.completed for obj in self.objectives.values()):
                self.complete()
                return True
        return False
        
    def start(self):
        """Start the quest."""
        if self.status == QuestStatus.NOT_STARTED:
            self.status = QuestStatus.IN_PROGRESS
            
    def complete(self):
        """Complete the quest."""
        if self.status == QuestStatus.IN_PROGRESS:
            self.status = QuestStatus.COMPLETED
            if self.on_complete:
                self.on_complete()
                
    def fail(self):
        """Fail the quest."""
        if self.status == QuestStatus.IN_PROGRESS:
            self.status = QuestStatus.FAILED
            if self.on_fail:
                self.on_fail()
                
    def is_completed(self) -> bool:
        """Check if quest is completed."""
        return self.status == QuestStatus.COMPLETED
        
    def get_progress(self) -> float:
        """Get overall quest progress as a percentage."""
        if not self.objectives:
            return 0.0
        return sum(obj.get_progress() for obj in self.objectives.values()) / len(self.objectives)

class QuestSystem:
    def __init__(self, quest_file: Optional[str] = None):
        self.quests: Dict[str, Quest] = {}
        self.active_quests: List[Quest] = []
        self.completed_quests: List[Quest] = []
        self.quest_log_visible = False
        if quest_file:
            self._load_quests(quest_file)
        
    def _load_quests(self, filepath: str):
        """Load quest data from JSON file."""
        try:
            with open(filepath, 'r') as f:
                quest_data = json.load(f)
                
            for quest_id, data in quest_data.items():
                quest = Quest(quest_id, data['title'], data['description'])
                
                # Add objectives
                for obj_id, obj_data in data.get('objectives', {}).items():
                    quest.add_objective(obj_id, obj_data['description'], 
                                     obj_data.get('target_amount', 1))
                    
                # Add rewards
                for reward_type, amount in data.get('rewards', {}).items():
                    quest.add_reward(reward_type, amount)
                    
                self.quests[quest_id] = quest
                
        except Exception as e:
            print(f"Error loading quest file: {e}")
            
    def start_quest(self, quest_id: str) -> bool:
        """Start a quest by its ID."""
        if quest_id not in self.quests:
            return False
            
        quest = self.quests[quest_id]
        if quest.status == QuestStatus.NOT_STARTED:
            quest.start()
            self.active_quests.append(quest)
            return True
        return False
        
    def complete_quest(self, quest_id: str) -> bool:
        """Complete a quest by its ID."""
        if quest_id not in self.quests:
            return False
            
        quest = self.quests[quest_id]
        if quest.status == QuestStatus.IN_PROGRESS:
            quest.complete()
            self.active_quests.remove(quest)
            self.completed_quests.append(quest)
            return True
        return False
        
    def fail_quest(self, quest_id: str) -> bool:
        """Fail a quest by its ID."""
        if quest_id not in self.quests:
            return False
            
        quest = self.quests[quest_id]
        if quest.status == QuestStatus.IN_PROGRESS:
            quest.fail()
            self.active_quests.remove(quest)
            return True
        return False
        
    def update_objective(self, quest_id: str, objective_id: str, amount: int = 1) -> bool:
        """Update progress of a quest objective."""
        if quest_id not in self.quests:
            return False
            
        return self.quests[quest_id].update_objective(objective_id, amount)
        
    def get_quest(self, quest_id: str) -> Optional[Quest]:
        """Get a quest by its ID."""
        return self.quests.get(quest_id)
        
    def get_active_quests(self) -> List[Quest]:
        """Get list of active quests."""
        return self.active_quests
        
    def get_completed_quests(self) -> List[Quest]:
        """Get list of completed quests."""
        return self.completed_quests
        
    def toggle_quest_log(self):
        """Toggle visibility of quest log."""
        self.quest_log_visible = not self.quest_log_visible
        
    def update(self, delta_time: float):
        """Atualiza o sistema de quests."""
        pass  # Por enquanto não precisamos atualizar nada
        
    def draw(self, screen: pygame.Surface):
        """Desenha o quest log."""
        if not self.quest_log_visible:
            return
            
        # Configurações do quest log
        padding = 20
        line_spacing = 10
        font = pygame.font.Font(None, 32)
        text_color = (255, 255, 255)
        title_color = (255, 255, 0)
        background_color = (0, 0, 0, 200)
        
        # Calcula dimensões
        screen_width = screen.get_width()
        screen_height = screen.get_height()
        log_width = min(400, screen_width - 2 * padding)
        
        # Prepara o texto
        lines = []
        lines.append(("QUEST LOG", title_color))
        lines.append(("", text_color))  # Espaço
        
        # Quests ativas
        lines.append(("Active Quests:", title_color))
        for quest in self.active_quests:
            lines.append((f"- {quest.title}", text_color))
            lines.append((f"  Progress: {quest.get_progress():.0%}", text_color))
            for obj in quest.objectives.values():
                lines.append((f"  • {obj.description}: {obj.current_amount}/{obj.target_amount}", 
                            text_color))
            lines.append(("", text_color))  # Espaço entre quests
            
        # Quests completadas
        if self.completed_quests:
            lines.append(("Completed Quests:", title_color))
            for quest in self.completed_quests:
                lines.append((f"- {quest.title}", text_color))
            
        # Calcula altura total
        log_height = len(lines) * (font.get_height() + line_spacing) + 2 * padding
        
        # Cria superfície do quest log com alpha
        log_surface = pygame.Surface((log_width, log_height), pygame.SRCALPHA)
        pygame.draw.rect(log_surface, background_color, (0, 0, log_width, log_height))
        
        # Desenha o texto
        y = padding
        for text, color in lines:
            if text:  # Não renderiza linhas vazias
                text_surface = font.render(text, True, color)
                log_surface.blit(text_surface, (padding, y))
            y += font.get_height() + line_spacing
            
        # Posiciona o quest log no canto superior direito
        log_x = screen_width - log_width - padding
        log_y = padding
        
        # Desenha na tela
        screen.blit(log_surface, (log_x, log_y))
